os/arch/arm/src/imxrt/imxrt_i2s.c

```diff
@@ -0,0 +1,2707 @@
+/****************************************************************************
+ *
+ * Copyright 2019 NXP Semiconductors All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
```

Comment:
Few comments on this file. Placing it here as it is applicable to more than one functions & dont want to duplicate across.

a. A single i2s device should be able to tranmsit and receive at the same time
b. i2s device should be able to be paused/stopped/resumed for only tx or only rx or both
c. i2s device should be able to operate in only tx mode, only rx mode & both ( and hence should be configurable for either or both)

Are above ensured in the code below?

Internally is i2s buffer a linear or pipelined?

Reply:
I add my comment to your questions

a. A single i2s device should be able to tranmsit and receive at the same time
 => I implemented i2s driver function by using SAI feature included in iMXRT series
       This feature is guaranteed by SAI feature 
b. i2s device should be able to be paused/stopped/resumed for only tx or only rx or both
 => I think it will work well if you properly call i2s_pause and i2s_resume in your side
c. i2s device should be able to operate in only tx mode, only rx mode & both ( and hence should be configurable for either or both)
 => From my implementation, there is no problem about it

As I said to you above, I plan to use imxrt_i2s_edma.c file including eDMA function.
So, I2S function has not any kind of buffer but DMA interface that can reduce cpu load

Reply:
In regard to your answer; "As I said to you above, I plan to use imxrt_i2s_edma.c file including eDMA function.
So, I2S function has not any kind of buffer but DMA interface that can reduce cpu load"

How is the life cycle of apb buffer in case of tx?
apb buffer would be filled by app and passed to i2s.

Is dma initiated with the same apb buffer? like address of buffer is the start address of dma and length same as length of the given buffer? or is a a copy in between?

Reply:
The initialization of eDMA includes the assignment of src. buffer and dest. buffer to DMA register.
Copy operation is executed in eDMA.
In our sdk program, there is no problem.
I think that this implementation needs to be checked